### PR TITLE
Finished Chapter 4 solutions

### PR DESCRIPTION
**Pull request by jacobprudhomme**
### Solutions for Chapter 4

I am looking for feedback on my instances of `Applicative` and `Monad` for `List` :smiley: Although the hint for the monad instance said I should make a helper that will flatten lists of lists, I found an implementation that doesn't require that, but I am not sure if it is correct.

Additionally, I am not sure if I implemented `andM` correctly. It is passing all the tests, but it seems a bit messy.

Thanks for the help!

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,10 +478,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure x = Reward x
```
Comment (chshersh) [inline src/Chapter4.hs]: Now, after you've mastered eta-reduction, you can apply such a technique to other functions as well  🙂 
But your implementation is already great 💯

File: src/Chapter4.hs [hunk 2]
```diff
@@ -472,10 +478,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
...
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Trap e) _ = Trap e
+    (<*>) (Reward f) x = fmap f x
```
Comment (chshersh) [inline src/Chapter4.hs:485]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 3]
```diff
@@ -488,7 +495,18 @@ Implement the 'Applicative' instance for our 'List' type.
   may also need to implement a few useful helper functions for our List
   type.
 -}
+appendLists :: List a -> List a -> List a
+appendLists Empty Empty = Empty
+appendLists Empty (Cons y ys) = Cons y (appendLists Empty ys)
```
Comment (chshersh) [inline src/Chapter4.hs]: These two cases can be combined into a single one. Note, that when the first list is `Empty`, you can just return the second list :slightly_smiling_face:

File: src/Chapter4.hs [hunk 4]
```diff
@@ -610,7 +629,10 @@ Implement the 'Monad' instance for our lists.
 🕯 HINT: You probably will need to implement a helper function (or
   maybe a few) to flatten lists of lists to a single list.
 -}
-
+instance Monad List where
+  (>>=) :: List a -> (a -> List b) -> List b
+  (>>=) Empty _ = Empty
+  (>>=) (Cons x xs) f = appendLists (f x) (xs >>= f)
```
Comment (chshersh) [inline src/Chapter4.hs]: This implementation is correct :+1: You can either pattern match manually and reuse `appendLists`, or you can implement the `flatten :: List (List a) -> List a` function and reuse `Functor` instance for `List`. Either way is fine, and they both have the same performance :slightly_smiling_face:

File: src/Chapter4.hs [hunk 5]
```diff
@@ -629,7 +651,7 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM b1 b2 = b1 >>= (\b -> if not b then pure b else pure (b &&) <*> b2)
```
Comment (chshersh) [inline src/Chapter4.hs]: Yes, your implementation is a bit overcomplicated, but don't worry! It is just a matter
```suggestion
andM b1 b2 = b1 >>= (\b -> if b then b2 else pure False)
```

File: src/Chapter4.hs [hunk 6]
```diff
@@ -672,7 +694,31 @@ Specifically,
...
+treeToPreorderList NoNode = []
+treeToPreorderList (Node x lt rt) = x : (treeToPreorderList lt ++ treeToPreorderList rt)
+
+treeToInorderList :: Tree a -> [a]
+treeToInorderList NoNode = []
+treeToInorderList (Node x lt rt) = treeToInorderList lt ++ [x] ++ treeToInorderList rt
+
+treeToPostorderList :: Tree a -> [a]
+treeToPostorderList NoNode = []
+treeToPostorderList (Node x lt rt) = treeToPostorderList lt ++ treeToPostorderList rt ++ [x]
```
Comment (chshersh) [inline src/Chapter4.hs:724]: Very nice! All three orders :ok_hand: :three:

**Review by chshersh — APPROVED** [review]
Great job on solving all four chapters!
And bonus points for cracking the advanced task :1st_place_medal: 

I left some comments, but they are just minor suggestions :slightly_smiling_face:

**Comment by jacobprudhomme** [comment]
@chshersh thank you for all the help! This was a really cool crash course on Haskell, and it felt very rewarding to do the work and get feedback on it. Best of luck, and a happy Hacktoberfest!

**Comment by chshersh** [comment]
@jacobprudhomme Thanks for your kind words about our course :blush: This means a lot to us!
We are happy that you enjoyed the project! :hugs: 

If you have a few minutes, we would appreciate your feedback on our course :slightly_smiling_face: (if you haven't submitted it yet)

* https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform

**Pull request merged**